README
Struct-Diff

The Struct::Diff module is a simple way to get structured diff for
nested perl structures (various mix of hashes of hashes and arrays of arrays)

INSTALLATION

To install this module, run the following commands:

    perl Makefile.PL
    make
    make test
    make install

SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the
perldoc command.

    perldoc Struct::Diff

You can also look for information at:

    RT, CPAN's request tracker (report bugs here)
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=Struct-Diff

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/Struct-Diff

    CPAN Ratings
        http://cpanratings.perl.org/d/Struct-Diff

    Search CPAN
        http://search.cpan.org/dist/Struct-Diff/


LICENSE AND COPYRIGHT

Copyright (C) 2015-2016 Michael Samoglyadov

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See L<http://dev.perl.org/licenses/> for more information.



Changes
Revision history for Struct-Diff

0.94    2017-12-25
        - 'freezer' opt to diff() added
        - path format in list_diff() corrected according Struct::Path v0.80

0.93    2017-12-09
        - diff(): get rid of redundant I tokens in array's diffs
        - diff(): get rid of empty subdiffs in array's diffs (noU)
        - patch(): references patching fixed
        - patch(): arrays patching fixed for LCS noU diffs
        - diff() and patch() code refactored
        - tests refactored and significantly improved
        - docs improved

0.92    2017-11-27
        - diff fixed for emptied/filled subhashes
        - patch fixed for cases when entire data type changed
        - subdiff squash fixed for A/R

0.91    2017-11-04
        - regexps comparison support added to diff()

0.90    2017-08-09
        - valid_diff() appeared
        - coderefs and blessed are now distinguishable for diff() by content
        - deprecated code removed
        - tests improved

0.89    2017-07-12
        - diff fixed for A/R hash subkeys with U subkeys present and noU opt
        - get rid of recursion in patch()
        - docs corrected, some cosmetic refactoring

0.88    2017-06-17
        - paths in list_diff() fixed for arrays in noU diffs
        - rename dsplit => split_diff(), old name is deprecated (alias)

0.87    2017-06-16
        - list_diff() added
        - get rid of global $Storable::canonical
        - dtraverse() deprecated (list_diff should be used instead)
        - code and docs refactorings

0.86    2017-01-22
        - number and same number as string are now distinguishable for diff()
        - diff() now much faster in hashes comparsion

0.85    2016-12-25
        - numbers stringification fixed in resulting diff

0.84    2016-12-23
        - arrays path indexes fixed in dtraverse() for noU diffs

0.83    2016-12-04
        - patch() fixed (A/R in the middle of arrays)

0.82    2016-10-21
        - docs corrected

0.81    2016-10-16
        - code and tests refactoring

0.80    2016-09-18
        - LCS for arrays diff implemented

0.71    2016-09-15
        - 'depth' opt to dtraverse() added

0.70    2016-09-13
        - dselect() removed, dtraverse() should be used instead
        - cosmetic refactoring
        - docs corrected

0.66    2016-09-06
        - ref to subdiff passed as fourth arg to callback functions in dtraverse()
        - dtraverse() expects true value as callback's out value

0.65    2016-08-16
        - tests fixed, min perl ver 5.008

0.64    2016-08-08
        - 'statuses' opt for dtraverse() appeared

0.63    2016-08-02
        - dtraverse() opts format changed, 'sortkeys' opt added

0.62    2016-07-15
        - docs corrected

0.61    2016-07-13
        - refactoring
        - diff()'s speed improved (~20%) on arrays

0.60    2016-07-11
        - dtraverse() appeared

0.58    2016-07-06
        - diff fixed for array's nested single items,
        - example: [[ 0 ]] vs [[ 1 ]], tests for this case added.

0.57    2016-07-03
        - unused code removed
        - more tests added, tests refactored

0.56    2016-06-22
        - pod and dist corrected, unused code removed.

0.55    2016-05-16
        - get rid of experimental feature (keys on scalars
        (not supported in Perl < 5.14 and forbidden since >= 5.20))

0.54    2016-05-16
        - dist corrected (tests and metadata).

0.53    2016-05-15
        - pod corrected.

0.52    2016-05-10
        - pod corrected, cosmetics.

0.50    2016-04-18
        - diff metadata format changed (and I hope stabilized)
        - diff()'s opts changed
        - strip() removed
        - dselect(), dsplit(), patch() appeared
        - tests and pod almost for everything

0.01    2015-07-21
        - initial release.



t/_common.pm
package _common;

# common parts for Struct::Path tests

use strict;
use warnings FATAL => 'all';
use parent qw(Exporter);

use Clone qw(clone);
use Data::Dumper qw();
use Struct::Diff qw(diff patch valid_diff);
use Test::More;

our @EXPORT = qw(
    run_batch_tests
    scmp
    sdump
);

my $EXP_KEYS;
if ($ENV{EXPORT_TESTS_DIR}) {

    $ENV{EXPORT_TESTS_DIR} =~ /(.*)/;
    $ENV{EXPORT_TESTS_DIR} = $1; # untaint

    $EXP_KEYS = { map { $_ => 1 } qw(a b diff opts) };

    require JSON;
}

sub run_batch_tests {
    local $Test::Builder::Level = $Test::Builder::Level + 1;

    for my $t (@_) {

        # export tests to JSON
        if ($EXP_KEYS and (!exists $t->{to_json} or $t->{to_json})) {
            die "Export dir doesn't exist" unless (-d $ENV{EXPORT_TESTS_DIR});

            my $data = clone($t);
            map { delete $data->{$_} unless ($EXP_KEYS->{$_}) } keys %{$data};

            my $file = "$ENV{EXPORT_TESTS_DIR}/$t->{name}.json";
            open(my $fh, '>', $file) or die "Failed to open file '$file' ($!)";
            print $fh JSON->new->pretty(1)->canonical(1)->encode($data);
            close($fh);
        }

        ### diff
        my $st = clone($t);
        my $diff = eval { diff($st->{a}, $st->{b}, %{$st->{opts}}) };
        if ($@) {
            fail("Diff: $@");
            fail("Patch: doesn't run");
            fail("Valid: doesn't run");
            next;
        }

        subtest "Diff " . $t->{name} => sub {
            is_deeply($diff, $st->{diff}, "Diff: $st->{name}") ||
                diag scmp($diff, $st->{diff});

            is_deeply($st->{a}, $t->{a}, "A mangled: $st->{name}") ||
                diag scmp($st->{a}, $t->{a});
            is_deeply($st->{b}, $t->{b}, "B mangled: $st->{name}") ||
                diag scmp($st->{b}, $t->{b});
        };

        ### patch
        if (
            !$t->{skip_patch}
            and not (
                $t->{opts}->{noA}
                or $t->{opts}->{noR}
                or $t->{opts}->{noN}
            )
        ) {
            my $st = clone($t);
            patch($st->{a}, $st->{diff});

            subtest "Patch " . $st->{name} => sub {
                is_deeply($st->{a}, $st->{b}, "Patch: $st->{name}") ||
                    diag scmp($st->{a}, $st->{b});

                is_deeply($t->{diff}, $st->{diff}, "Patch: diff mangled: $st->{name}") ||
                    diag scmp($t->{diff}, $st->{diff});
            };
        } else {
            pass("patch skipped");
        }

        ### valid_diff
        $diff = clone($t->{diff});

        subtest "Valid " . $t->{name} => sub {
            ok(valid_diff($diff));

            is_deeply($diff, $t->{diff}, "Valid: diff mangled: $t->{name}") ||
                diag scmp($diff, $t->{diff});
        };
    }
}

sub scmp($$) {
    return "GOT: " . sdump(shift) . ";\nEXP: " . sdump(shift) . ";";
}

sub sdump($) {
    return Data::Dumper->new([shift])->Terse(1)->Sortkeys(1)->Quotekeys(0)->Indent(0)->Deepcopy(1)->Dump();
}

1;



examples.pl
#!/usr/bin/env perl

use strict;
use warnings FATAL => 'all';

use lib 'lib';

use Data::Dumper;
use Struct::Diff qw(diff dselect dsplit patch);

$Data::Dumper::Indent = 0;
$Data::Dumper::Quotekeys = 0;
$Data::Dumper::Sortkeys = 1;

my @examples = (
    {
        'a' => "Some old value",
        'b' => "Some new value",
        'desc' => "Scalars",
        'opts_diff' => {},
    },
    {
        'a' => [ 2, 3, 4, 5, 6, 7 ],
        'b' => [ 2, 3, 9, 5, 6, 7 ],
        'desc' => "Arrays, one lement changed",
        'opts_diff' => {},
    },
    {
        'a' => [ 2, 3, 4, 5, 6, 7, 8 ],
        'b' => [ 2, 3, 4, 5, 6, 7 ],
        'desc' => "Arrays, one element removed",
        'opts_diff' => {},
        'opts_dselect' => { states => { R => 1 }, fromD => [] },
    },
    {
        'a' => [ 2, 3, 4, 5, 6, 7, 8 ],
        'b' => [ 2, 3, 4, 5, 6, 7 ],
        'desc' => "Same, without common",
        'opts_diff' => { noU => 1 },
    },
    {
        'a' => { x => { y => { z => 'v' }}},
        'b' => { x => { y => { z => 'cv' }}},
        'desc' => "Hashes, deep-down val changed",
        'opts_diff' => {},
    },
    {
        'a' => { x => [ 7, { y => 4 } ] },
        'b' => { x => [ 7, { y => 9 } ], z => 33 },
        'desc' => "Mixed (ARRAYS/HASHES) structure",
        'opts_diff' => {},
    },
    {
        'a' => { x => [ 7, { y => 4 } ] },
        'b' => { x => [ 7, { y => 9 } ], z => 33 },
        'desc' => "Same, but without common items and old values",
        'opts_diff' => { noU => 1, noO => 1 },
        'opts_dselect' => { fromD => [ 'z' ]},
        'run_dsplit' => 1,
    },
);

print "\n";

for my $e (@examples) {
    my $d = diff($e->{a}, $e->{b}, %{$e->{opts_diff}});
    print "####### $e->{desc} #######\n";
    print Data::Dumper->Dump([$e->{a}], ["a"]), "\n";
    print Data::Dumper->Dump([$e->{b}], ["b"]), "\n\n";

    print Data::Dumper->Dump([$e->{opts_diff}], ["*opts_diff"]), "\n";
    {
#        local $Data::Dumper::Indent = 2;
        print Data::Dumper->Dump([$d], ["diff"]), "\n\n";
    }
    if ($e->{opts_dselect}) {
        my @s = dselect($d, %{$e->{opts_dselect}});
        print Data::Dumper->Dump([$e->{opts_dselect}], ["*opts_dselect"]), "\n";
        print Data::Dumper->Dump([\@s], ["dselect"]), "\n\n";
    }
    if ($e->{run_dsplit}) {
        my $s = dsplit($d);
        for my $k ('a', 'b') {
            if (exists $s->{$k}) {
                print Data::Dumper->Dump([$s->{$k}], ["dsplit->{$k}"]), "\n\n";
            } else {
                print "\$dsplit->{$k} not exists\n";
            }
        }
    }
}


lib/Struct/Diff.pm
package Struct::Diff;

use 5.006;
use strict;
use warnings FATAL => 'all';
use parent qw(Exporter);

use Scalar::Util qw(looks_like_number);
use Storable 2.05 qw(freeze);
use Algorithm::Diff qw(sdiff);

our @EXPORT_OK = qw(
    diff
    list_diff
    patch
    split_diff
    valid_diff
);

=head1 NAME

Struct::Diff - Recursive diff for nested perl structures

=begin html

<a href="https://travis-ci.org/mr-mixas/Struct-Diff.pm"><img src="https://travis-ci.org/mr-mixas/Struct-Diff.pm.svg?branch=master" alt="Travis CI"></a>
<a href='https://coveralls.io/github/mr-mixas/Struct-Diff.pm?branch=master'><img src='https://coveralls.io/repos/github/mr-mixas/Struct-Diff.pm/badge.svg?branch=master' alt='Coverage Status'/></a>
<a href="https://badge.fury.io/pl/Struct-Diff"><img src="https://badge.fury.io/pl/Struct-Diff.svg" alt="CPAN version"></a>

=end html

=head1 VERSION

Version 0.94

=cut

our $VERSION = '0.94';

=head1 SYNOPSIS

    use Struct::Diff qw(diff list_diff split_diff patch valid_diff);

    $x = {one => [1,{two => 2}]};
    $y = {one => [1,{two => 9}],three => 3};

    $diff = diff($x, $y, noO => 1, noU => 1); # omit unchanged items and old values
    # $diff == {D => {one => {D => [{D => {two => {N => 9}},I => 1}]},three => {A => 3}}}

    @list_diff = list_diff($diff); # list (path and ref pairs) all diff entries
    # @list_diff == ({keys => ['one']},[1],{keys => ['two']}],\{N => 9},[{keys => ['three']}],\{A => 3})

    $splitted = split_diff($diff);
    # $splitted->{a} # does not exist
    # $splitted->{b} == {one => [{two => 9}],three => 3}

    patch($x, $diff); # $x now equal to $y by structure and data

    @errors = valid_diff($diff);

=head1 EXPORT

Nothing is exported by default.

=head1 DIFF FORMAT

Diff is simply a HASH whose keys shows status for each item in passed
structures. Every status type (except C<D>) may be omitted during the diff
calculation. Disabling some or other types produce different diffs: diff with
only unchanged items is also possible (when all other types disabled).

=over 4

=item A

Stands for 'added' (exists only in second structure), it's value - added item.

=item D

Means 'different' and contains subdiff. The only status type which can't be
disabled.

=item I

Index for array item, used only when prior item was omitted.

=item N

Is a new value for changed item.

=item O

Alike C<N>, C<O> is a changed item's old value.

=item R

Similar for C<A>, but for removed items.

=item U

Represent unchanged items.

=back

Diff format: metadata alternates with data and, as a result, diff may represent
any structure of any data types. Simple types specified as is, arrays and hashes
contain subdiffs for their items with native for such types addressing: indexes
for arrays and keys for hashes.

Sample:

    old:  {one => [5,7]}
    new:  {one => [5],two => 2}
    opts: {noU => 1} # omit unchanged items

    diff:
    {D => {one => {D => [{I => 1,R => 7}]},two => {A => 2}}}
    ||    | |     ||    |||    | |    |     |     ||    |
    ||    | |     ||    |||    | |    |     |     ||    +- with value 2
    ||    | |     ||    |||    | |    |     |     |+- it says key was added
    ||    | |     ||    |||    | |    |     |     +- subdiff for it
    ||    | |     ||    |||    | |    |     +- another key from top-level hash
    ||    | |     ||    |||    | |    +- what it was (item's value: 7)
    ||    | |     ||    |||    | +- shows what happened to item (removed)
    ||    | |     ||    |||    +- array item's actual index
    ||    | |     ||    ||+- prior item was omitted
    ||    | |     ||    |+- subdiff for array item
    ||    | |     ||    +- it's value - ARRAY
    ||    | |     |+- it is deeply changed
    ||    | |     +- subdiff for key 'one'
    ||    | +- it has key 'one'
    ||    +- top-level thing is a HASH
    |+- changes somewhere deeply inside
    +- diff is always a HASH

=head1 SUBROUTINES

=head2 diff

Returns hashref to recursive diff between two passed things. Beware when
changing diff: it's parts are links to original structures.

    $diff  = diff($x, $y, %opts);
    $patch = diff($x, $y, noU => 1, noO => 1, trimR => 1); # smallest possible diff

=head3 Options

=over 4

=item freezer C<< <sub> >>

Serializer callback (redefines default serializer). L<Storable/freeze> is used
by default, see L</CONFIGURATION VARIABLES> for details.

=item noX C<< <true|false> >>

Where X is a status (C<A>, C<N>, C<O>, C<R>, C<U>); such status will be omitted.

=item trimR C<< <true|false> >>

Drop removed item's data.

=back

=cut

our $Freezer = sub {
    local $Storable::canonical = 1; # for equal snapshots for equal by data hashes
    local $Storable::Deparse = 1;   # for coderefs

    freeze \$_[0];
};

sub diff($$;@) {
    my ($x, $y, %opts) = @_;

    $opts{freezer} = $Freezer unless (exists $opts{freezer});

    _diff($x, $y, %opts);
}

sub _diff($$;@);
sub _diff($$;@) {
    my ($x, $y, %opts) = @_;

    my $d = {};
    my $type = ref $x;

    if ($type ne ref $y) {
        $d->{O} = $x unless ($opts{noO});
        $d->{N} = $y unless ($opts{noN});
    } elsif ($type eq 'ARRAY' and $x != $y) {
        return $opts{noU} ? {} : { U => [] } unless (@{$x} or @{$y});

        my ($i, $I) = (-1, -1);
        for (sdiff($x, $y, $opts{freezer})) {
            $i++;

            if ($_->[0] eq 'u') {
                unless ($opts{noU}) {
                    if (exists $d->{D}) {
                        push @{$d->{D}}, { U => $_->[1] };
                    } else { # nobody else - fill U version
                        push @{$d->{U}}, $_->[1];
                    }
                }
                next;
            } elsif (exists $d->{U}) { # diff should be converted to D type
                map { push @{$d->{D}}, { U => $_ } } @{delete $d->{U}};
            }

            if ($_->[0] eq 'c') {
                my $sd = _diff($_->[1], $_->[2], %opts);
                push @{$d->{D}}, $sd if (keys %{$sd});
            } elsif ($_->[0] eq '+') {
                push @{$d->{D}}, { A => $_->[2] } unless ($opts{noA});
            } else { # '-'
                push @{$d->{D}}, { R => $opts{trimR} ? undef : $_->[1] }
                    unless ($opts{noR});
            }

            $d->{D}->[-1]->{I} = $I = $i
                if (exists $d->{D} and $#{$d->{D}} != $i and ++$I != $i);
        }
    } elsif ($type eq 'HASH' and $x != $y) {
        my @keys = keys %{{ %{$x}, %{$y} }}; # uniq keys for both hashes
        return $opts{noU} ? {} : { U => {} } unless (@keys);

        for my $k (@keys) {
            if (exists $x->{$k} and exists $y->{$k}) {
                if ($opts{freezer}($x->{$k}) eq $opts{freezer}($y->{$k})) {
                    $d->{U}->{$k} = $y->{$k} unless ($opts{noU});
                } else {
                    my $sd = _diff($x->{$k}, $y->{$k}, %opts);
                    $d->{D}->{$k} = $sd if (keys %{$sd});
                }
                next;
            }

            if (exists $x->{$k}) {
                $d->{D}->{$k}->{R} = $opts{trimR} ? undef : $x->{$k}
                    unless ($opts{noR});
            } else {
                $d->{D}->{$k}->{A} = $y->{$k} unless ($opts{noA});
            }
        }

        if (exists $d->{U} and exists $d->{D}) {
            map { $d->{D}->{$_}->{U} = $d->{U}->{$_} } keys %{$d->{U}};
            delete $d->{U};
        }
    } elsif ($type && $x == $y || $opts{freezer}($x) eq $opts{freezer}($y)) {
        $d->{U} = $x unless ($opts{noU});
    } else {
        $d->{O} = $x unless ($opts{noO});
        $d->{N} = $y unless ($opts{noN});
    }

    return $d;
}

=head2 list_diff

List pairs (path_to_subdiff, ref_to_subdiff)) for provided diff. See
L<Struct::Path/ADDRESSING SCHEME> for path format specification.

    @list = list_diff($diff);

=head3 Options

=over 4

=item depth C<< <int> >>

Don't dive deeper than defined number of levels. C<undef> used by default
(unlimited).

=item sort C<< <sub|true|false> >>

Defines how to handle hash subdiffs. Keys will be picked randomely (default
C<keys> behavior), sorted by provided subroutine (if value is a coderef) or
lexically sorted if set to some other true value.

=back

=cut

sub list_diff($;@) {
    my @stack = ([], \shift); # init: (path, diff)
    my %opts = @_;
    my ($diff, @list, $path);

    while (@stack) {
        ($path, $diff) = splice @stack, 0, 2;

        if (!exists ${$diff}->{D} or $opts{depth} and @{$path} >= $opts{depth}) {
            push @list, $path, $diff;
        } elsif (ref ${$diff}->{D} eq 'ARRAY') {
            map {
                unshift @stack,
                    [@{$path},
                        [exists ${$diff}->{D}->[$_]->{I}
                            ? ${$diff}->{D}->[$_]->{I} # use provided index
                            : $_
                        ]
                    ],
                    \${$diff}->{D}->[$_]
            } reverse 0 .. $#{${$diff}->{D}};
        } else { # HASH
            map {
                unshift @stack, [@{$path}, {K => [$_]}], \${$diff}->{D}->{$_}
            } $opts{sort}
                ? ref $opts{sort} eq 'CODE'
                    ? reverse $opts{sort}(keys %{${$diff}->{D}})
                    : reverse sort keys %{${$diff}->{D}}
                : keys %{${$diff}->{D}};
        }
    }

    return @list;
}

=head2 split_diff

Divide diff to pseudo original structures.

    $structs = split_diff(diff($x, $y));
    # $structs->{a}: items from $x
    # $structs->{b}: items from $y

=cut

sub split_diff($);
sub split_diff($) {
    my $d = $_[0];
    my (%out, $sd);

    if (exists $d->{D}) {
        if (ref $d->{D} eq 'ARRAY') {
            for (@{$d->{D}}) {
                $sd = split_diff($_);
                push @{$out{a}}, $sd->{a} if (exists $sd->{a});
                push @{$out{b}}, $sd->{b} if (exists $sd->{b});
            }
        } else { # HASH
            for (keys %{$d->{D}}) {
                $sd = split_diff($d->{D}->{$_});
                $out{a}->{$_} = $sd->{a} if (exists $sd->{a});
                $out{b}->{$_} = $sd->{b} if (exists $sd->{b});
            }
        }
    } elsif (exists $d->{U}) {
        $out{a} = $out{b} = $d->{U};
    } elsif (exists $d->{A}) {
        $out{b} = $d->{A};
    } elsif (exists $d->{R}) {
        $out{a} = $d->{R};
    } else {
        $out{b} = $d->{N} if (exists $d->{N});
        $out{a} = $d->{O} if (exists $d->{O});
    }

    return \%out;
}

=head2 patch

Apply diff.

    patch($target, $diff);

=cut

sub patch($$) {
    my @stack = (\$_[0], $_[1]); # ref to alias - to be able to change passed scalar

    while (@stack) {
        my ($s, $d) = splice @stack, 0, 2; # struct, subdiff

        if (exists $d->{D}) {
            if (ref $d->{D} eq 'ARRAY') {
                my ($i, $r) = (0, 0); # struct array idx, removed items counter

                for (@{$d->{D}}) {
                    $i = $_->{I} - $r if (exists $_->{I});

                    if (exists $_->{D} or exists $_->{N}) {
                        push @stack, \${$s}->[$i], $_;
                    } elsif (exists $_->{A}) {
                        splice @{${$s}}, $i, 1,
                            (@{${$s}} > $i ? ($_->{A}, ${$s}->[$i]) : $_->{A});
                    } elsif (exists $_->{R}) {
                        splice @{${$s}}, $i, 1;
                        $r++;
                        next; # don't increment $i
                    }

                    $i++;
                }
            } else { # HASH
                while (my ($k, $v) = each %{$d->{D}}) {
                    if (exists $v->{D} or exists $v->{N}) {
                        push @stack, \${$s}->{$k}, $v;
                    } elsif (exists $v->{A}) {
                        ${$s}->{$k} = $v->{A};
                    } elsif (exists $v->{R}) {
                        delete ${$s}->{$k};
                    }
                }
            }
        } elsif (exists $d->{N}) {
            ${$s} = $d->{N};
        }
    }
}

=head2 valid_diff

Validate diff structure. In scalar context returns C<1> for valid diff, C<undef>
otherwise. In list context returns list of pairs (path, type) for each error. See
L<Struct::Path/ADDRESSING SCHEME> for path format specification.

    @errors_list = valid_diff($diff); # list context

or

    $is_valid = valid_diff($diff); # scalar context

=cut

sub valid_diff($) {
    my @stack = ([], shift); # (path, diff)
    my ($diff, @errs, $path);

    while (@stack) {
        ($path, $diff) = splice @stack, 0, 2;

        unless (ref $diff eq 'HASH') {
            return undef unless wantarray;
            push @errs, $path, 'BAD_DIFF_TYPE';
            next;
        }

        if (exists $diff->{D}) {
            if (ref $diff->{D} eq 'ARRAY') {
                map {
                    unshift @stack, [@{$path}, [$_]], $diff->{D}->[$_]
                } 0 .. $#{$diff->{D}};
            } elsif (ref $diff->{D} eq 'HASH') {
                map {
                    unshift @stack, [@{$path}, {K => [$_]}], $diff->{D}->{$_}
                } sort keys %{$diff->{D}};
            } else {
                return undef unless wantarray;
                unshift @errs, $path, 'BAD_D_TYPE';
            }
        }

        if (exists $diff->{I}) {
            if (!looks_like_number($diff->{I}) or int($diff->{I}) != $diff->{I}) {
                return undef unless wantarray;
                unshift @errs, $path, 'BAD_I_TYPE';
            }

            if (keys %{$diff} < 2) {
                return undef unless wantarray;
                unshift @errs, $path, 'LONESOME_I';
            }
        }
    }

    return wantarray ? @errs : 1;
}

=head1 CONFIGURATION VARIABLES

=over 4

=item $Struct::Diff::Freezer

Contains reference to default serialization function (C<diff()> rely on it
to determine data equivalency). L<Storable/freeze> with enabled
C<$Storable::canonical> and C<$Storable::Deparse> opts used by default.

L<Data::Dumper> is suitable for structures containing regular experrions:

    $Struct::Diff::Freezer = sub {
        local $Data::Dumper::Deparse    = 1;
        local $Data::Dumper::Sortkeys   = 1;
        local $Data::Dumper::Terse      = 1;

        return Dumper @_;
    }

But, comparing to C<Storable> it has two another issues: speed and unability
to distinguish numbers from their string representations.

=back

=head1 LIMITATIONS

Only arrays and hashes traversed. All other data types compared by reference
addresses and content.

L<Storable/freeze> (serializer used by default) failes on compiled regexp
serialization, so, consider to use other serializer if data contains regular
expressions. See L<CONFIGURATION VARIABLES> for details.

Struct::Diff fails on structures with loops in references. C<has_circular_ref>
from L<Data::Structure::Util> can help to detect such structures.

=head1 AUTHOR

Michael Samoglyadov, C<< <mixas at cpan.org> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-struct-diff at rt.cpan.org>,
or through the web interface at
L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Struct-Diff>. I will be notified,
and then you'll automatically be notified of progress on your bug as I make
changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Struct::Diff

You can also look for information at:

=over 4

=item * RT: CPAN's request tracker (report bugs here)

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Struct-Diff>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Struct-Diff>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Struct-Diff>

=item * Search CPAN

L<http://search.cpan.org/dist/Struct-Diff/>

=back

=head1 SEE ALSO

L<Algorithm::Diff>, L<Data::Deep>, L<Data::Diff>, L<Data::Difference>,
L<JSON::MergePatch>

L<Data::Structure::Util>, L<Struct::Path>, L<Struct::Path::PerlStyle>

=head1 LICENSE AND COPYRIGHT

Copyright 2015-2017 Michael Samoglyadov.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See L<http://dev.perl.org/licenses/> for more information.

=cut

1; # End of Struct::Diff


benchmark/mk_struct.pl
#!/usr/bin/env perl

use strict;
use warnings;

use Data::Dumper;

# recursive struct maker
sub smk($$;$);
sub smk($$;$) {
    my $depth = shift || return (rand(1) > 0.5) ? rand(65535) : int(rand(65535));
    my $width = shift;
    my $style = shift;

    my $out;
    if ((defined $style and ref $style eq 'ARRAY') or (not defined $style and rand(1) > 0.5)) {
        map { push @{$out}, smk($depth - 1, $width, $style) } 0..($width - 1);
    } else {
        map { $out->{$_} = smk($depth - 1, $width, $style) }
            map { (rand(1) > 0.5) ? rand(65535) : int(rand(65535)) } 0..($width - 1);
    }
    return $out;
}

srand(time + $$);

my $result;

$result->{'AoA'} = smk(6, 5, []);
$result->{'HoH'} = smk(6, 5, {});
$result->{'MIX'} = smk(6, 5);

$Data::Dumper::Indent = 0;
$Data::Dumper::Sortkeys = 1;
$Data::Dumper::Varname = "STRUCT";

print "package Structs;\n";
print "our ", Dumper($result), "\n";
print "1;\n"


Makefile.PL
use 5.006;
use strict;
use warnings FATAL => 'all';
use ExtUtils::MakeMaker;

WriteMakefile(
    NAME             => 'Struct::Diff',
    AUTHOR           => 'Michael Samoglyadov <mixas@cpan.org>',
    VERSION_FROM     => 'lib/Struct/Diff.pm',
    ABSTRACT_FROM    => 'lib/Struct/Diff.pm',
    LICENSE          => 'Perl_5',
    PL_FILES         => {},
    MIN_PERL_VERSION => 5.008,
    CONFIGURE_REQUIRES => {
        'ExtUtils::MakeMaker' => 0,
    },
    BUILD_REQUIRES => {
        'Data::Dumper'    => 0,
        'Test::More'      => 0,
    },
    META_MERGE => {
        resources => {
            repository  => 'https://github.com/mr-mixas/Struct-Diff.pm',
        },
    },
    PREREQ_PM => {
        'Algorithm::Diff' => 1.19,
        'Carp'            => 0,
        'Exporter'        => 0,
        'Storable'        => 2.05,
        'parent'          => 0,
        'strict'          => 0,
        'warnings'        => 0,
    },
    dist  => {
        COMPRESS => 'gzip -9f',
        SUFFIX   => 'gz',
    },
    clean => { FILES => 'Struct-Diff-*' },
);
